### PR TITLE
Add a schema inference test

### PR DESCRIPTION
**Pull request by rodesai**
This patch adds a test for schema inference. The test ensures that
schemas are inferred correctly and that the correct statement is
produced.

To facillitate the test, the code for inference was moved into a new
class called StatementWithInferredSchema. StatementWithInferredSchema
has two members, the statement and the statement text, and a static
function for inferring the schema from a statement that maybe doesn't
have a schema.

This patch also includes fixes for some issues found along the way
    - Don't fail statements for which the backing topic has a schema
      with types that are not supported by KSQL. Instead, log an error
      and omit those fields from the schema.
    - Bug in the schema translator for handling null names
    - Bug in SqlFormatter for handling statements with empty schemas

### Description 
_What behavior do you want to change, why, how does your patch achieve the changes?_

### Testing done 
_Describe the testing stategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java [hunk 1]
```diff
@@ -240,8 +237,9 @@ StructuredDataSource getResultDatasource(final Select select, final String name)
     );
   }
 
-  private AbstractStreamCreateStatement maybeAddFieldsFromSchemaRegistry(
-      AbstractStreamCreateStatement streamCreateStatement
+  private StatementWithInferredSchema maybeAddFieldsFromSchemaRegistry(
+      AbstractStreamCreateStatement streamCreateStatement,
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java]: can we make this `final` and not re-assign it in the method? i.e., return from line 269

File: ksql-engine/src/main/java/io/confluent/ksql/util/StatementWithInferredSchema.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+package io.confluent.ksql.util;
+
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.tree.AbstractStreamCreateStatement;
+import io.confluent.ksql.parser.tree.Statement;
+
+import java.util.Map;
+
+public class StatementWithInferredSchema {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/util/StatementWithInferredSchema.java]: maybe just `StatementWithSchema` as it isn't always inferred

File: ksql-engine/src/test/java/io/confluent/ksql/util/AvroSchemaInferenceTest.java [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.HashMap;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+public class AvroSchemaInferenceTest {
+  private Schema getSchemaForDdlStatement(final AbstractStreamCreateStatement statement) {
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/util/AvroSchemaInferenceTest.java]: nit: maybe move private methods to the bottom?

File: ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectSchemaTranslator.java [hunk 4]
```diff
@@ -20,21 +20,43 @@
...
+    final KsqlException nonStructRootError = new KsqlException(
+        "KSQL stream/table schema must be structured");
+    try {
+      final Schema rowSchema = toKsqlFieldSchema(schema);
+      if (rowSchema.type() != Schema.Type.STRUCT) {
+        throw nonStructRootError;
+      }
+      return rowSchema;
+    } catch (UnsupportedTypeException e) {
+      log.error("Unsupported type at root of schema: %s", e.getMessage());
```
Comment (dguy) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectSchemaTranslator.java]: log and throw? Maybe we should just throw a `KsqlException` with the appropriate message and cause?

File: ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java [hunk 5]
```diff
@@ -268,15 +266,20 @@ private AbstractStreamCreateStatement maybeAddFieldsFromSchemaRegistry(
...
           )
       );
-      streamCreateStatement = streamCreateStatement.copyWith(
+      final AbstractStreamCreateStatement statementWithProperties = streamCreateStatement.copyWith(
           streamCreateStatement.getElements(),
-          newProperties
+          newProperties);
+      return StatementWithSchema.forStatement(
+          statementWithProperties,
+          SqlFormatter.formatSql(statementWithProperties),
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java]: One thing to note for the future: the test coverage for `SqlFormatter` is very sparse. We rely heavily on this code to serialize SQL statements, and we should really expand coverage for this class at some point. 

One way to do this is to expand the query translation test to execute the re-serialized version of the statement in addition to the regular version to ensure that the result is the same. 

Does that make sense? If so, would you mind filing an issue with that suggestion. We can then fix it going forward as we focus on improving our test coverage.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/SqlFormatter.java [hunk 6]
```diff
@@ -347,8 +347,11 @@ protected Void visitCreateStream(CreateStream node, Integer indent) {
               .append(" ")
               .append(tableElement.getType());
         }
-        builder.append(")").append(" WITH (");
-        addComma = false;
+        builder.append(")");
+      }
+      if (!node.getProperties().isEmpty()) {
```
Comment (apurvam) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/SqlFormatter.java]: Is this a regression or a fix for an existing bug?

Comment (rodesai) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/SqlFormatter.java]: This is a fix for an existing bug.

File: ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java (hunk 5 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java]: Filed https://github.com/confluentinc/ksql/issues/1571

**Pull request merged**